### DISCUSSION TITLE
Determining appropriate configuration for our use case

### DISCUSSION THREAD

**Discussion by ArjunNamineni** [quartznet/quartznet#1576] [Q&A]
Hi there,

We've implemented Quartz in a windows service in some of our test environments and are running some performance tests with varying results. We've been playing around with some configuration changes to see how it impacts performance for our use case.

For some background, we have 3 nodes that do the create and delete of triggers/jobs. And 3 separate nodes that run StdScheduler. Our typical load is creating around 600-900 triggers per minute; most of with are scheduled to be fired 5 minutes into the future, and nearly all within 3 hours. About 1/4-1/3 (130-300) triggers are deleted by the client before it fires if it's no longer needed. And most jobs takes around 15ms to complete, almost all less than 50ms. So our use case is many, short running jobs.

We create a new job + trigger every time, each one fires only once, and is removed (ie no recurring/calendar triggers or jobs).

We are currently running Quartz 3.0.7 and ado.net/sql server job store. The following is our configuration:

```
quartz.threadPool.type - Quartz.Simpl.SimpleThreadPool, Quartz
quartz.threadPool.threadCount - 10
quartz.threadPool.threadPriority - Normal
quartz.jobStore.misfireThreshold - 60000
quartz.jobStore.type - Quartz.Impl.AdoJobStore.JobStoreTX, Quartz
quartz.jobStore.useProperties - false
quartz.jobStore.clustered - true
quartz.jobStore.acquireTriggersWithinLock - false
quartz.jobStore.dataSource - default
quartz.jobStore.tablePrefix - QRTZ_
quartz.jobStore.driverDelegateType - Quartz.Impl.AdoJobStore.SqlServerDelegate, Quartz
quartz.serializer.type - json
quartz.scheduler.instanceName - QuartzScheduler
quartz.scheduler.instanceId - AUTO
quartz.scheduler.batchTriggerAcquisitionMaxCount - 1
quartz.scheduler.batchTriggerAcquisitionFireAheadTimeWindow - 0
```

We've been running some performance tests to see how Quartz performs when there are many triggers in the QRTZ_TRIGGERS table. So we created 2 million triggers and ran load tests of creating/firing/deleting triggers and jobs at the rate described above. But we've run into issues of queries to QRTZ_LOCKS timing out consistently as there seem to be a long blocking chain of requests waiting for the row lock for the record in QRTZ_LOCKS. This is causing slowness in firing events at the throughput we're expecting.

The sql call duration to other tables (ie QRTZ_TRIGGERS) also jumps from sub 5ms to over 50ms during the tests, but the total volume of triggers to be fired also increases. So we are unsure if the issue is sql performance, or our current Quartz configuration is not optimized for our use case and not able to process fast enough.

I'd like to test with some changes to the configuration, and am curious if I can get some help understanding what would make sense for our use case. Mainly, I'd like to try changes to `threadCount`, `batchTriggerAcquisitionMaxCount`, using multiple schedulers.

So a few questions:

1. I was thinking that reducing `threadCount` would reduce the number of sql calls and maybe the number of blocking calls to QRTZ_LOCKS, but that didn't seem to have any effect on performance. Does `threadCount` control the number of threads that are trying to acquire triggers? Or the number of threads available to execute jobs? Would reducing `threadCount` help us? What metrics should we look at to determine the appropriate `threadCount` to use?

2. I was also thinking about setting 'acquireTriggersWithinLock' to 'true', using 'batchTriggerAcquisitionMaxCount' of '10', and 'batchTriggerAcquisitionFireAheadTimeWindow' of '1000'. In the hopes that there would be less frequent need to acquire db lock on QRTZ_LOCKS. Would our use case benefit from batch trigger acquisition? What metrics should we look at to determine the appropriate 'batchTriggerAcquisitionMaxCount'?

3. Lastly, I'm thinking we could benefit from multiple schedulers so that not all jobs need to wait for the same db lock, since all jobs can be executed in parallel. Would that improve the performance issues we are seeing? What metrics should we look at to determine the appropriate number of schedulers to use? We want to be careful to not create too many since it may be hard to reduce number of schedulers without creating orphan triggers.

I know there's a lot of questions there. We're really excited to try Quartz out and want to make sure we're using it to it's fullest potential. Thank you very much!

**Comment by lahma** [comment]
First of all, thank you for opening this discussion which such great level of details. It's always a lot easier to start digesting the problem when background is clear(er).

> We are currently running Quartz 3.0.7 and ado.net/sql server job store. The following is our configuration:

There has been some performance improvements done, like to use some joins which reduces network round-trips and using fixed parameter sizes to cache query plans.

Please consider upgrading as first step if that's possible. 

You should check the [detailed release notes for 3.1](https://github.com/quartznet/quartznet/releases/tag/v3.1.0) and later. Please note that the indexes have been updated and you should re-run the index creation script which drops unnecessary ones thus reducing index update pressure.

> We've been running some performance tests to see how Quartz performs when there are many triggers in the QRTZ_TRIGGERS table. So we created 2 million triggers and ran load tests of creating/firing/deleting triggers and jobs at the rate described above. But we've run into issues of queries to QRTZ_LOCKS timing out consistently as there seem to be a long blocking chain of requests waiting for the row lock for the record in QRTZ_LOCKS. This is causing slowness in firing events at the throughput we're expecting.

2 million triggers is A LOT. If they are being fired at the same time, it's going to create a lot of contention against the locking table (distributed lock). The more there are concurrent trigger fires, the more contention the table will get. If you have 2 million triggers which are evenly distributed they won't be fed into triggering queue causing so much locking - so make sure you model your stress test properly.

> The sql call duration to other tables (ie QRTZ_TRIGGERS) also jumps from sub 5ms to over 50ms during the tests, but the total volume of triggers to be fired also increases. So we are unsure if the issue is sql performance, or our current Quartz configuration is not optimized for our use case and not able to process fast enough.

With SQL Server delegate you should have quite optimal locking mechanism. But please note that there have been more fixes after the release you are currently evaluating.

> 1. I was thinking that reducing `threadCount` would reduce the number of sql calls and maybe the number of blocking calls to QRTZ_LOCKS, but that didn't seem to have any effect on performance. Does `threadCount` control the number of threads that are trying to acquire triggers? Or the number of threads available to execute jobs? Would reducing `threadCount` help us? What metrics should we look at to determine the appropriate `threadCount` to use?

Thread count is just basically a number of concurrent triggers allowed on a node. It will most likely just slow you down with low value and high acquire count.

> 2. I was also thinking about setting 'acquireTriggersWithinLock' to 'true', using 'batchTriggerAcquisitionMaxCount' of '10', and 'batchTriggerAcquisitionFireAheadTimeWindow' of '1000'. In the hopes that there would be less frequent need to acquire db lock on QRTZ_LOCKS. Would our use case benefit from batch trigger acquisition? What metrics should we look at to determine the appropriate 'batchTriggerAcquisitionMaxCount'?

This is something you need to test. But under a lot of load it makes sense to have larger value for `batchTriggerAcquisitionMaxCount` and enable the `acquireTriggersWithinLock`. Batch acquisition count should follow the thread count (actually concurrent task count). For modern hardware 10 is quite low, you should test different values against your environment.

> 3. Lastly, I'm thinking we could benefit from multiple schedulers so that not all jobs need to wait for the same db lock, since all jobs can be executed in parallel. Would that improve the performance issues we are seeing? What metrics should we look at to determine the appropriate number of schedulers to use? We want to be careful to not create too many since it may be hard to reduce number of schedulers without creating orphan triggers.

You can of course split to multiple schedulers if they are logically separate. But if you still use the same DB the load might not be affected in positive way.

If possible, please try the latest version and run the schema/index update script - it should be enough to update the indexes, the bottom part of the script.

**Reply by lahma** [reply]
And also please check [the configuration reference](https://www.quartz-scheduler.net/documentation/quartz-3.x/configuration/reference.html) if you haven't already, it has some explanations.

**Reply by ArjunNamineni** [reply]
Thanks for the speedy response!

I did notice that 3.1 removed and re-ordered some indices, so will try the upgrade.

> 2 million triggers is A LOT. If they are being fired at the same time, it's going to create a lot of contention against the locking table

Sorry, I should have been more clear about the details of the performance test is doing. We created 2 million triggers to fire years in the future. Then started creating around 10 triggers a second to be fired in 5 minutes, for about 6-7 hours. This was to simulate how the sql calls would perform as the table grew larger. Not necessarily with 2 million triggers to fire at the same time.

We've noticed that if we see a very high number of triggers being created, sql call duration will increase, leading to longer times in acquiring the distributed lock. This leads to a slower throughput and the number of triggers in the table increasing. If it reaches around 2 million triggers (due to the reduced throughput/high traffic), we start seeing QRTZ_LOCKS timeouts and even moderate load (ie even after high load reduces) causes misfires.

> Batch acquisition count should follow the thread count (actually concurrent task count)

What is concurrent task count? Is that the [quartz.threadPool.maxConcurrency](https://www.quartz-scheduler.net/documentation/quartz-3.x/configuration/reference.html#threadpool)? maxConcurrency is a feature of Quartz 3.1 correct?

> For modern hardware 10 is quite low, you should test different values against your environment.

Sounds good, we'll test with some higher values.

> And also please check [the configuration reference](https://www.quartz-scheduler.net/documentation/quartz-3.x/configuration/reference.html) if you haven't already, it has some explanations.

There are some details in the [clustering section](https://www.quartz-scheduler.net/documentation/quartz-3.x/configuration/reference.html#clustering) that suggests "If you need to scale out to support thousands of short-running (e.g 1 second) jobs, consider partitioning the set of jobs by using multiple distinct schedulers". Would that guidance apply to our use case since we create hundreds of triggers per minute and each job takes less than 50ms?

Personally, I feel like we have the most to gain from using multiple schedulers so that each thread is not waiting on the distributed lock (each job execution is logically separated, so all can run in parallel). Since we are seeing timeouts in acquiring the lock, would this help with that? Or you think our issues would be from database performance of the other sql calls?

Although we would use the same DB instance for each scheduler, I think we might benefit from partitioning triggers in filenames (filegroups?) based on the SchedulerName.

**Reply by lahma** [reply]
> > 2 million triggers is A LOT. If they are being fired at the same time, it's going to create a lot of contention against the locking table
> 
> Sorry, I should have been more clear about the details of the performance test is doing. We created 2 million triggers to fire years in the future. Then started creating around 10 triggers a second to be fired in 5 minutes, for about 6-7 hours. This was to simulate how the sql calls would perform as the table grew larger. Not necessarily with 2 million triggers to fire at the same time.
> 
> We've noticed that if we see a very high number of triggers being created, sql call duration will increase, leading to longer times in acquiring the distributed lock. This leads to a slower throughput and the number of triggers in the table increasing. If it reaches around 2 million triggers (due to the reduced throughput/high traffic), we start seeing QRTZ_LOCKS timeouts and even moderate load (ie even after high load reduces) causes misfires.

It would be nice if you could share the sample application creating the load. Basically taking single trigger under distributed lock is always wasteful compared to acquiring multiple to run.

> > Batch acquisition count should follow the thread count (actually concurrent task count)
> 
> What is concurrent task count? Is that the [quartz.threadPool.maxConcurrency](https://www.quartz-scheduler.net/documentation/quartz-3.x/configuration/reference.html#threadpool)? maxConcurrency is a feature of Quartz 3.1 correct?

`maxConcurrency` is a term used in the new default thread pool of Quartz 3 where tasks are run using shared CLR task system. There's also dedicated thread pool if true separate threads are wanted, but it won't help with the SQL pressure.

> There are some details in the [clustering section](https://www.quartz-scheduler.net/documentation/quartz-3.x/configuration/reference.html#clustering) that suggests "If you need to scale out to support thousands of short-running (e.g 1 second) jobs, consider partitioning the set of jobs by using multiple distinct schedulers". Would that guidance apply to our use case since we create hundreds of triggers per minute and each job takes less than 50ms?
> 
> Personally, I feel like we have the most to gain from using multiple schedulers so that each thread is not waiting on the distributed lock (each job execution is logically separated, so all can run in parallel). Since we are seeing timeouts in acquiring the lock, would this help with that? Or you think our issues would be from database performance of the other sql calls?
> 
> Although we would use the same DB instance for each scheduler, I think we might benefit from partitioning triggers in filenames (filegroups?) based on the SchedulerName.

I think multiple schedulers is something you should investigate and test for your use case. Of course it would be ideal if we could tune Quartz to perform well under you scenario without splitting to multiple schedulers until required. Batch acquisition is the key here to find a balance with three node system and taking enough of the queue reducing lock infrastructure load.
